File: CS/Models/DataObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.Data;
using System.Collections;
using System.ComponentModel;
using System.Linq.Expressions;
using DevExpress.Data.Filtering;
using System.Reflection;
using Example.Models;

public class DataObject : IListServer, ITypedList {
    DataClassesDataContext context = new DataClassesDataContext();

    readonly IQueryable<Contact> origQuery;

    IQueryable<Contact> query;
    String keyExpression;

    const Int32 pageSize = 23; // approximate page size

    // cache accessibility
    Boolean isCountDirty = true;
    Boolean isResultDirty = true;
    Boolean isKeysDirty = true;

    // cached values
    Int32 count = -1; // row count
    Array keys;  // array of keys
    Array rows;  // array of records
    Int32 startKeyIndex = -1; // relative key index
    Int32 startRowIndex = -1; // relative row index

    public DataObject() {
        this.origQuery = from contact in context.Contacts
                         select contact;

        this.query = this.origQuery;
        this.keyExpression = "ContactID";
    }

    void IListServer.Apply(DevExpress.Data.Filtering.CriteriaOperator filterCriteria, ICollection<ServerModeOrderDescriptor> sortInfo, int groupCount, ICollection<ServerModeSummaryDescriptor> groupSummaryInfo, ICollection<ServerModeSummaryDescriptor> totalSummaryInfo) {
        isCountDirty = true;
        isResultDirty = true;
        isKeysDirty = true;

        count = -1;
        startKeyIndex = -1;
        startRowIndex = -1;

        this.query = this.origQuery;

        if (sortInfo != null) {
            Boolean first = true;
            IOrderedQueryable<Contact> oq = null;

            foreach (ServerModeOrderDescriptor info in sortInfo) {
                String propertyName = (info.SortExpression as OperandProperty).PropertyName;

                if (first) {
                    if (info.IsDesc)
                        oq = LinqHelper.OrderByDescending<Contact>(this.query, propertyName);
                    else
                        oq = LinqHelper.OrderBy<Contact>(this.query, propertyName);

                    first = false;
                }
                else {
                    if (info.IsDesc)
                        oq = LinqHelper.ThenByDescending<Contact>(oq, propertyName);
                    else
                        oq = LinqHelper.ThenBy<Contact>(oq, propertyName);
                }
            }

            this.query = oq.AsQueryable<Contact>();
        }
    }

    event EventHandler<ServerModeExceptionThrownEventArgs> IListServer.ExceptionThrown {
        add { throw new NotImplementedException(); }
        remove { throw new NotImplementedException(); }
    }

    int IListServer.FindIncremental(DevExpress.Data.Filtering.CriteriaOperator expression, string value, int startIndex, bool searchUp, bool ignoreStartRow, bool allowLoop) {
        throw new NotImplementedException();
    }

    IList IListServer.GetAllFilteredAndSortedRows() {
        throw new NotImplementedException();
    }

    List<ListSourceGroupInfo> IListServer.GetGroupInfo(ListSourceGroupInfo parentGroup) {
        throw new NotImplementedException();
    }

    int IListServer.GetRowIndexByKey(object key) {
        throw new NotImplementedException();
    }

    object IListServer.GetRowKey(int index) {
        if (isKeysDirty ||
            (index < startKeyIndex) ||
            (index >= startKeyIndex + pageSize)) {
            isKeysDirty = false;
            startKeyIndex = index;

            var keysQuery = from obj in query.Skip(index).Take(pageSize)
                            select obj.GetType().GetProperty(this.keyExpression).GetValue(obj, null);
            var keysArray = keysQuery.ToArray();

            this.keys = Array.CreateInstance(keysQuery.ElementType, keysArray.Count());

            Array.Copy(keysArray, this.keys, keysArray.Count());
        }

        return this.keys.GetValue(index - startKeyIndex);
    }

    List<object> IListServer.GetTotalSummary() {
        throw new NotImplementedException();
    }

    object[] IListServer.GetUniqueColumnValues(DevExpress.Data.Filtering.CriteriaOperator expression, int maxCount, bool includeFilteredOut) {
        throw new NotImplementedException();
    }

    event EventHandler<ServerModeInconsistencyDetectedEventArgs> IListServer.InconsistencyDetected {
        add { throw new NotImplementedException(); }
        remove { throw new NotImplementedException(); }
    }

    int IListServer.LocateByValue(DevExpress.Data.Filtering.CriteriaOperator expression, object value, int startIndex, bool searchUp) {
        throw new NotImplementedException();
    }

    void IListServer.Refresh() {
        throw new NotImplementedException();
    }

    int IList.Add(object value) {
        throw new NotImplementedException();
    }

    void IList.Clear() {
        throw new NotImplementedException();
    }

    bool IList.Contains(object value) {
        throw new NotImplementedException();
    }

    int IList.IndexOf(object value) {
        throw new NotImplementedException();
    }

    void IList.Insert(int index, object value) {
        throw new NotImplementedException();
    }

    bool IList.IsFixedSize {
        get { throw new NotImplementedException(); }
    }

    bool IList.IsReadOnly {
        get { throw new NotImplementedException(); }
    }

    void IList.Remove(object value) {
        throw new NotImplementedException();
    }

    void IList.RemoveAt(int index) {
        throw new NotImplementedException();
    }

    object IList.this[int index] {
        get {
            if (isResultDirty ||
                (index < startRowIndex) ||
                (index >= startRowIndex + pageSize)) {
                isResultDirty = false;
                startRowIndex = index;

                var rowsQuery = from obj in query.Skip(index).Take(pageSize)
                                select obj;
                var rowsArray = rowsQuery.ToArray();

                this.rows = Array.CreateInstance(rowsQuery.ElementType, rowsArray.Count());

                Array.Copy(rowsArray, this.rows, rowsArray.Count());
            }

            return this.rows.GetValue(index - startRowIndex);
        }
        set {
            throw new NotImplementedException();
        }
    }

    void ICollection.CopyTo(Array array, int index) {
        throw new NotImplementedException();
    }

    int ICollection.Count {
        get {
            if (isCountDirty) {
                isCountDirty = false;
                count = query.Count();
            }

            return count;
        }
    }

    bool ICollection.IsSynchronized {
        get { throw new NotImplementedException(); }
    }

    object ICollection.SyncRoot {
        get { throw new NotImplementedException(); }
    }

    IEnumerator IEnumerable.GetEnumerator() {
        throw new NotImplementedException();
    }

    PropertyDescriptorCollection ITypedList.GetItemProperties(PropertyDescriptor[] listAccessors) {
        return TypeDescriptor.GetProperties(query.ElementType);
    }

    string ITypedList.GetListName(PropertyDescriptor[] listAccessors) {
        return query.ElementType.Name;
    }



}

/* http://stackoverflow.com/questions/41244/dynamic-linq-orderby */

public static class LinqHelper {
    public static IOrderedQueryable<T> OrderBy<T>(this IQueryable<T> source, string property) {
        return ApplyOrder<T>(source, property, "OrderBy");
    }

    public static IOrderedQueryable<T> OrderByDescending<T>(this IQueryable<T> source, string property) {
        return ApplyOrder<T>(source, property, "OrderByDescending");
    }

    public static IOrderedQueryable<T> ThenBy<T>(this IOrderedQueryable<T> source, string property) {
        return ApplyOrder<T>(source, property, "ThenBy");
    }

    public static IOrderedQueryable<T> ThenByDescending<T>(this IOrderedQueryable<T> source, string property) {
        return ApplyOrder<T>(source, property, "ThenByDescending");
    }

    static IOrderedQueryable<T> ApplyOrder<T>(IQueryable<T> source, string property, string methodName) {
        string[] props = property.Split('.');
        Type type = typeof(T);
        ParameterExpression arg = Expression.Parameter(type, "x");
        Expression expr = arg;
        foreach (string prop in props) {
            // use reflection (not ComponentModel) to mirror LINQ
            PropertyInfo pi = type.GetProperty(prop);
            expr = Expression.Property(expr, pi);
            type = pi.PropertyType;
        }
        Type delegateType = typeof(Func<,>).MakeGenericType(typeof(T), type);
        LambdaExpression lambda = Expression.Lambda(delegateType, expr, arg);

        object result = typeof(Queryable).GetMethods().Single(
                method => method.Name == methodName
                        && method.IsGenericMethodDefinition
                        && method.GetGenericArguments().Length == 2
                        && method.GetParameters().Length == 2)
                .MakeGenericMethod(typeof(T), type)
                .Invoke(null, new object[] { source, lambda });
        return (IOrderedQueryable<T>)result;
    }
}

File: VB/Models/DataObject.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports DevExpress.Data
Imports System.Collections
Imports System.ComponentModel
Imports System.Linq.Expressions
Imports DevExpress.Data.Filtering
Imports System.Reflection

Public Class DataObject
    Implements IListServer
    Implements ITypedList
    Private context As New DataClassesDataContext()

    ReadOnly origQuery As IQueryable(Of Contact)

    Private query As IQueryable(Of Contact)
    Private keyExpression As [String]

    Const pageSize As Int32 = 23
    ' approximate page size
    ' cache accessibility
    Private isCountDirty As [Boolean] = True
    Private isResultDirty As [Boolean] = True
    Private isKeysDirty As [Boolean] = True

    ' cached values
    Private count As Int32 = -1
    ' row count
    Private keys As Array
    ' array of keys
    Private rows As Array
    ' array of records
    Private startKeyIndex As Int32 = -1
    ' relative key index
    Private startRowIndex As Int32 = -1
    ' relative row index
    Public Sub New()
        Me.origQuery = From contact In context.Contacts Select contact

        Me.query = Me.origQuery
        Me.keyExpression = "ContactID"
    End Sub

    Private Sub IListServer_Apply(ByVal filterCriteria As DevExpress.Data.Filtering.CriteriaOperator, ByVal sortInfo As ICollection(Of ServerModeOrderDescriptor), ByVal groupCount As Integer, ByVal groupSummaryInfo As ICollection(Of ServerModeSummaryDescriptor), ByVal totalSummaryInfo As ICollection(Of ServerModeSummaryDescriptor)) Implements IListServer.Apply
        isCountDirty = True
        isResultDirty = True
        isKeysDirty = True

        count = -1
        startKeyIndex = -1
        startRowIndex = -1

        Me.query = Me.origQuery

        If sortInfo IsNot Nothing Then
            Dim first As [Boolean] = True
            Dim oq As IOrderedQueryable(Of Contact) = Nothing

            For Each info As ServerModeOrderDescriptor In sortInfo
                Dim propertyName As [String] = TryCast(info.SortExpression, OperandProperty).PropertyName

                If first Then
                    If info.IsDesc Then
                        oq = LinqHelper.OrderByDescending(Of Contact)(Me.query, propertyName)
                    Else
                        oq = LinqHelper.OrderBy(Of Contact)(Me.query, propertyName)
                    End If

                    first = False
                Else
                    If info.IsDesc Then
                        oq = LinqHelper.ThenByDescending(Of Contact)(oq, propertyName)
                    Else
                        oq = LinqHelper.ThenBy(Of Contact)(oq, propertyName)
                    End If
                End If
            Next

            Me.query = oq.AsQueryable()
        End If
    End Sub

    Private Custom Event ExceptionThrown As EventHandler(Of ServerModeExceptionThrownEventArgs) Implements IListServer.ExceptionThrown
        AddHandler(ByVal value As EventHandler(Of ServerModeExceptionThrownEventArgs))
            Throw New NotImplementedException()
        End AddHandler
        RemoveHandler(ByVal value As EventHandler(Of ServerModeExceptionThrownEventArgs))
            Throw New NotImplementedException()
        End RemoveHandler
        RaiseEvent(ByVal sender As System.Object, ByVal e As ServerModeExceptionThrownEventArgs)
        End RaiseEvent
    End Event

    Private Function IListServer_FindIncremental(ByVal expression As DevExpress.Data.Filtering.CriteriaOperator, ByVal value As String, ByVal startIndex As Integer, ByVal searchUp As Boolean, ByVal ignoreStartRow As Boolean, ByVal allowLoop As Boolean) As Integer Implements IListServer.FindIncremental
        Throw New NotImplementedException()
    End Function

    Private Function IListServer_GetAllFilteredAndSortedRows() As IList Implements IListServer.GetAllFilteredAndSortedRows
        Throw New NotImplementedException()
    End Function

    Private Function IListServer_GetGroupInfo(ByVal parentGroup As ListSourceGroupInfo) As List(Of ListSourceGroupInfo) Implements IListServer.GetGroupInfo
        Throw New NotImplementedException()
    End Function

    Private Function IListServer_GetRowIndexByKey(ByVal key As Object) As Integer Implements IListServer.GetRowIndexByKey
        Throw New NotImplementedException()
    End Function

    Private Function IListServer_GetRowKey(ByVal index As Integer) As Object Implements IListServer.GetRowKey
        If isKeysDirty OrElse (index < startKeyIndex) OrElse (index >= startKeyIndex + pageSize) Then
            isKeysDirty = False
            startKeyIndex = index

            Dim keysQuery = From obj In query.Skip(index).Take(pageSize) _
                           Select obj.GetType().GetProperty(Me.keyExpression).GetValue(obj, Nothing)
            Dim keysArray = keysQuery.ToArray()

            Me.keys = Array.CreateInstance(keysQuery.ElementType, keysArray.Count())

            Array.Copy(keysArray, Me.keys, keysArray.Count())
        End If

        Return Me.keys.GetValue(index - startKeyIndex)
    End Function

    Private Function IListServer_GetTotalSummary() As Dictionary(Of Object, Object) Implements IListServer.GetTotalSummary
        Throw New NotImplementedException()
    End Function

    Private Function IListServer_GetUniqueColumnValues(ByVal expression As DevExpress.Data.Filtering.CriteriaOperator, ByVal maxCount As Integer, ByVal includeFilteredOut As Boolean) As Object() Implements IListServer.GetUniqueColumnValues
        Throw New NotImplementedException()
    End Function

    Private Custom Event InconsistencyDetected As EventHandler(Of ServerModeInconsistencyDetectedEventArgs) Implements IListServer.InconsistencyDetected
        AddHandler(ByVal value As EventHandler(Of ServerModeInconsistencyDetectedEventArgs))
            Throw New NotImplementedException()
        End AddHandler
        RemoveHandler(ByVal value As EventHandler(Of ServerModeInconsistencyDetectedEventArgs))
            Throw New NotImplementedException()
        End RemoveHandler
        RaiseEvent(ByVal sender As System.Object, ByVal e As ServerModeInconsistencyDetectedEventArgs)
        End RaiseEvent
    End Event

    Private Function IListServer_LocateByValue(ByVal expression As DevExpress.Data.Filtering.CriteriaOperator, ByVal value As Object, ByVal startIndex As Integer, ByVal searchUp As Boolean) As Integer Implements IListServer.LocateByValue
        Throw New NotImplementedException()
    End Function

    Private Sub IListServer_Refresh() Implements IListServer.Refresh
        Throw New NotImplementedException()
    End Sub

    Private Function IList_Add(ByVal value As Object) As Integer Implements IList.Add
        Throw New NotImplementedException()
    End Function

    Private Sub IList_Clear() Implements IList.Clear
        Throw New NotImplementedException()
    End Sub

    Private Function IList_Contains(ByVal value As Object) As Boolean Implements IList.Contains
        Throw New NotImplementedException()
    End Function

    Private Function IList_IndexOf(ByVal value As Object) As Integer Implements IList.IndexOf
        Throw New NotImplementedException()
    End Function

    Private Sub IList_Insert(ByVal index As Integer, ByVal value As Object) Implements IList.Insert
        Throw New NotImplementedException()
    End Sub

    Private ReadOnly Property IList_IsFixedSize() As Boolean Implements IList.IsFixedSize
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Private ReadOnly Property IList_IsReadOnly() As Boolean Implements IList.IsReadOnly
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Private Sub IList_Remove(ByVal value As Object) Implements IList.Remove
        Throw New NotImplementedException()
    End Sub

    Private Sub IList_RemoveAt(ByVal index As Integer) Implements IList.RemoveAt
        Throw New NotImplementedException()
    End Sub

    Public Property IList_Item(ByVal index As Integer) As Object Implements IList.Item
        Get
            If isResultDirty OrElse (index < startRowIndex) OrElse (index >= startRowIndex + pageSize) Then
                isResultDirty = False
                startRowIndex = index

                Dim rowsQuery = From obj In query.Skip(index).Take(pageSize) _
                                 Select obj
                Dim rowsArray = rowsQuery.ToArray()

                Me.rows = Array.CreateInstance(rowsQuery.ElementType, rowsArray.Count())

                Array.Copy(rowsArray, Me.rows, rowsArray.Count())
            End If

            Return Me.rows.GetValue(index - startRowIndex)
        End Get
        Set(ByVal value As Object)
            Throw New NotImplementedException()
        End Set
    End Property

    Private Sub ICollection_CopyTo(ByVal array As Array, ByVal index As Integer) Implements ICollection.CopyTo
        Throw New NotImplementedException()
    End Sub

    Private ReadOnly Property ICollection_Count() As Integer Implements ICollection.Count
        Get
            If isCountDirty Then
                isCountDirty = False
                count = query.Count()
            End If

            Return count
        End Get
    End Property

    Private ReadOnly Property ICollection_IsSynchronized() As Boolean Implements ICollection.IsSynchronized
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Private ReadOnly Property ICollection_SyncRoot() As Object Implements ICollection.SyncRoot
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
        Throw New NotImplementedException()
    End Function

    Private Function ITypedList_GetItemProperties(ByVal listAccessors As PropertyDescriptor()) As PropertyDescriptorCollection Implements ITypedList.GetItemProperties
        Return TypeDescriptor.GetProperties(query.ElementType)
    End Function

    Private Function ITypedList_GetListName(ByVal listAccessors As PropertyDescriptor()) As String Implements ITypedList.GetListName
        Return query.ElementType.Name
    End Function



End Class